crates/ibc/src/applications/transfer/context.rs

```diff
@@ -9,23 +9,26 @@ use crate::applications::transfer::relay::on_recv_packet::process_recv_packet;
 use crate::applications::transfer::relay::on_timeout_packet::process_timeout_packet;
 use crate::applications::transfer::{PrefixedCoin, PrefixedDenom, VERSION};
 use crate::core::ics04_channel::channel::{Counterparty, Order};
-use crate::core::ics04_channel::context::{ChannelKeeper, ChannelReader};
+use crate::core::ics04_channel::commitment::PacketCommitment;
+use crate::core::ics04_channel::context::{ChannelKeeper, SendPacketReader};
+use crate::core::ics04_channel::error::Error as Ics04Error;
+use crate::core::ics04_channel::handler::send_packet::SendPacketResult;
 use crate::core::ics04_channel::handler::ModuleExtras;
 use crate::core::ics04_channel::msgs::acknowledgement::Acknowledgement as GenericAcknowledgement;
-use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics04_channel::packet::{Packet, Sequence};
 use crate::core::ics04_channel::Version;
 use crate::core::ics24_host::identifier::{ChannelId, ConnectionId, PortId};
 use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
 use crate::prelude::*;
 use crate::signer::Signer;
 
 pub trait Ics20Keeper:
-    ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
+    Ics20ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
 {
     type AccountId;
```

Comment:
How about we just merge `Ics20ChannelKeeper` in here, remove this associated type?

Reply:
I prefer keeping it this way because it makes it easier to specify the blanket impl (i.e. `impl<T> Ics20ChannelKeeper for T`).

Reply:
This seems to work:

```diff
diff --git a/crates/ibc/src/applications/transfer/context.rs b/crates/ibc/src/applications/transfer/context.rs
index 16ea001..2da1ba6 100644
--- a/crates/ibc/src/applications/transfer/context.rs
+++ b/crates/ibc/src/applications/transfer/context.rs
@@ -22,39 +22,7 @@ use crate::core::ics26_routing::context::{ModuleOutputBuilder, OnRecvPacketAck};
 use crate::prelude::*;
 use crate::signer::Signer;
 
-pub trait Ics20Keeper:
-    Ics20ChannelKeeper + BankKeeper<AccountId = <Self as Ics20Keeper>::AccountId>
-{
-    type AccountId;
-}
-
-pub trait Ics20Reader: SendPacketReader {
-    type AccountId: TryFrom<Signer>;
-
-    /// get_port returns the portID for the transfer module.
-    fn get_port(&self) -> Result<PortId, Ics20Error>;
-
-    /// Returns the escrow account id for a port and channel combination
-    fn get_channel_escrow_address(
-        &self,
-        port_id: &PortId,
-        channel_id: &ChannelId,
-    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error>;
-
-    /// Returns true iff send is enabled.
-    fn is_send_enabled(&self) -> bool;
-
-    /// Returns true iff receive is enabled.
-    fn is_receive_enabled(&self) -> bool;
-
-    /// Returns a hash of the prefixed denom.
-    /// Implement only if the host chain supports hashed denominations.
-    fn denom_hash_string(&self, _denom: &PrefixedDenom) -> Option<String> {
-        None
-    }
-}
-
-pub trait Ics20ChannelKeeper {
+pub trait Ics20Keeper: BankKeeper {
     fn store_send_packet_result(&mut self, result: SendPacketResult) -> Result<(), Ics04Error> {
         self.store_next_sequence_send(
             result.port_id.clone(),
@@ -87,9 +55,35 @@ pub trait Ics20ChannelKeeper {
     ) -> Result<(), Ics04Error>;
 }
 
-impl<T> Ics20ChannelKeeper for T
+pub trait Ics20Reader: SendPacketReader {
+    type AccountId: TryFrom<Signer>;
+
+    /// get_port returns the portID for the transfer module.
+    fn get_port(&self) -> Result<PortId, Ics20Error>;
+
+    /// Returns the escrow account id for a port and channel combination
+    fn get_channel_escrow_address(
+        &self,
+        port_id: &PortId,
+        channel_id: &ChannelId,
+    ) -> Result<<Self as Ics20Reader>::AccountId, Ics20Error>;
+
+    /// Returns true iff send is enabled.
+    fn is_send_enabled(&self) -> bool;
+
+    /// Returns true iff receive is enabled.
+    fn is_receive_enabled(&self) -> bool;
+
+    /// Returns a hash of the prefixed denom.
+    /// Implement only if the host chain supports hashed denominations.
+    fn denom_hash_string(&self, _denom: &PrefixedDenom) -> Option<String> {
+        None
+    }
+}
+
+impl<T> Ics20Keeper for T
 where
-    T: ChannelKeeper,
+    T: ChannelKeeper + BankKeeper,
 {
     fn store_packet_commitment(
         &mut self,
```

Reply:
https://github.com/cosmos/ibc-rs/pull/244/commits/ccff7f7a4675b35de91c18a8e8a1533d181009b6